torch/nn/modules/module.py

```diff
@@ -597,76 +611,109 @@ def state_dict(self, destination=None, prefix='', keep_vars=False):
+                    input_name = input_name.split('.', 1)[0]  # get the name of param/buffer/child
+                    if input_name not in self._modules and input_name not in local_state:
+                        unexpected_keys.append(key)
 
     def load_state_dict(self, state_dict, strict=True):
         r"""Copies parameters and buffers from :attr:`state_dict` into
-        this module and its descendants. If :attr:`strict` is ``True`` then
+        this module and its descendants. If :attr:`strict` is ``True``, then
         the keys of :attr:`state_dict` must exactly match the keys returned
-        by this module's :func:`state_dict()` function.
+        by this module's :meth:`~torch.nn.Module.state_dict` function.
 
         Arguments:
-            state_dict (dict): A dict containing parameters and
+            state_dict (dict): a dict containing parameters and
                 persistent buffers.
-            strict (bool): Strictly enforce that the keys in :attr:`state_dict`
-                match the keys returned by this module's `:func:`state_dict()`
-                function.
+            strict (bool, optional): whether to strictly enforce that the keys
+                in :attr:`state_dict` match the keys returned by this module's
+                :meth:`~torch.nn.Module.state_dict` function. Default: ``True``
         """
-        def submodule_key_mismatch(full_name, is_missing):
-            module = self
-            names = full_name.split(".")
-            for module_name in names[:-1]:
-                if module_name in module._modules:
-                    module = module._modules[module_name]
-                else:
-                    return
-            module._load_state_dict_key_mismatch(full_name, names[-1], is_missing)
-
-        unexpected = []
-        own_state = self.state_dict()
-        for name, param in state_dict.items():
-            if name in own_state:
-                if isinstance(param, Parameter):
-                    # backwards compatibility for serialized parameters
-                    param = param.data
-                try:
-                    own_state[name].copy_(param)
-                except Exception:
-                    raise RuntimeError('While copying the parameter named {}, '
-                                       'whose dimensions in the model are {} and '
-                                       'whose dimensions in the checkpoint are {}.'
-                                       .format(name, own_state[name].size(), param.size()))
-            elif strict:
-                unexpected.append(name)
+        missing_keys = []
+        unexpected_keys = []
+        error_msgs = []
+
+        # copy state_dict so _load_from_state_dict can modify it
+        metadata = getattr(state_dict, '_metadata', None)
+        state_dict = state_dict.copy()
+        if metadata is not None:
+            state_dict._metadata = metadata
+
+        def load(module, prefix=''):
```

Comment:
`load` here does not necessarily have to be a local func def. However, that `load_state_dict` is not a recursive call predates this patch. If I were to guess, I would say that this probably made the original implementation easier. For custom modules, you can, for now, overwrite `_load_from_state_dict` instead. However, note that it is a private API at the moment.

Reply:
@SsnL Thanks for your quick response.
The problem is that by overriding `_load_from_state_dict` you can't change the behaviour of how you treat children modules.
Say my custom Module has certain children Modules and I don't want to serialize/deserialize the values of the parameters of all the models (save only modules those are not blacklisted in that custom module). Here I can't override the behaviour of traversal, as it is fixed to the top module.
It would be great to have the ability to override the recursive traversal code.
The potential usage is saving/loading of very compact models on top of huge pretrained models. By having control over the loading behaviour we can drastically reduce footprint of checkpoints.

Reply:
@SsnL May I consider creating Pull Request to make `load` a method of `Module`?